File: activity/20200805/RabbitLee/Symmetric_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if root is None:
            return True
        return self.isSymmetricTrees(root.left, root.right)
        
    def isSymmetricTrees(self, p: TreeNode, q: TreeNode) -> bool:
        """
        detect whether p and q are symmetirc
        """
        if p and q:
            return p.val == q.val and self.isSymmetricTrees(p.left, q.right) \
                and self.isSymmetricTrees(p.right, q.left)
        return p is q

File: activity/20200805/scau-flyingbird/Symmetric_tree.go
/**
 * Definition for a binary tree node.
  * type TreeNode struct {
	   *     Val int
	    *     Left *TreeNode
	     *     Right *TreeNode
	      * }
*/

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	} else {
		return isSymmetricT(root.Left, root.Right)
	}
}

func isSymmetricT(left *TreeNode, right *TreeNode) bool {
	if left == nil && right == nil {
		return true
	} else if left == nil || right == nil {
		return false
	} else if left.Val != right.Val {
		return false
	} else {
		return isSymmetricT(left.Left, right.Right) && isSymmetricT(left.Right, right.Left)
	}

}
